corehq/motech/dhis2/entities_helpers.py

```diff
@@ -112,15 +113,25 @@ def find_tracked_entity_instances(requests, case_trigger_info, case_config):
     return finder.find_tracked_entity_instances(case_trigger_info)
 
 
-def update_tracked_entity_instance(tracked_entity, requests, case_trigger_info, case_config):
+def update_tracked_entity_instance(requests, tracked_entity, etag, case_trigger_info, case_config):
     for attr_id, value_source in case_config.attributes.items():
         set_te_attr(tracked_entity, attr_id, value_source.get_value(case_trigger_info))
     enrollments = get_enrollments(case_trigger_info, case_config)
     if enrollments:
         tracked_entity["enrollments"] = enrollments
     tei_id = tracked_entity["trackedEntityInstance"]
     endpoint = f"/api/{DHIS2_API_VERSION}/trackedEntityInstances/{tei_id}"
-    response = requests.put(endpoint, json=tracked_entity, raise_for_status=True)
+    headers = {
+        "Content-type": "application/json",
+        "Accept": "application/json"
+    }
+    if not etag.startswith("W/"):  # weak ETags never match
```

Comment:
In what situations are "weak ETags" used? I'm unfamiliar with that concept and curious about how it could affect the semantics here. If they are used frequently then it means the optimistic concurrency controls will be disabled and the chances of race conditions may be higher.

Reply:
Weak etags are cheap to generate, and are used for [caching](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#Weak_validation). But apparently they are not useful for OCC / preventing "[mid-air collisions](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag#Avoiding_mid-air_collisions)". I will test this to see if DHIS2 returns 412 if we pass an "If-Match" header with a weak etag.